### PR TITLE
Contracts refactor. Bye, truffle!!! :clap:

### PR DESCRIPTION
**Pull request by LogvinovLeon**
This PR:

* Removes truffle
* Makes tests independent from each other
* Uses our deployer to migrate needed contracts in tests
* Moves a lot of types/utils around because now they're shared between 0x.js and contracts
* Prepares contracts to use abi-gen (doesn't use it yet)

### CODE REVIEW

**Comment by abandeali1** [comment]
Is there a reason that `contracts/test/ts/utils` is different from `contracts/util`? Can we combine these?

File: packages/contracts/package.json [hunk 1]
```diff
@@ -44,11 +47,12 @@
         "copyfiles": "^1.2.0",
         "dirty-chai": "^2.0.1",
         "mocha": "^4.0.1",
+        "npm-run-all": "^4.1.2",
         "solc": "^0.4.18",
         "truffle": "^4.0.1",
```
Comment (abandeali1) [inline packages/contracts/package.json]: Forgot the most important part :)

File: packages/contracts/test/exchange/helpers.ts [hunk 2]
```diff
@@ -1,52 +1,68 @@
...
 
     before(async () => {
-        const [tokenRegistry, exchange] = await Promise.all([TokenRegistry.deployed(), Exchange.deployed()]);
-        exchangeWrapper = new ExchangeWrapper(exchange);
-        const [repAddress, dgdAddress] = await Promise.all([
-            tokenRegistry.getTokenAddressBySymbol('REP'),
-            tokenRegistry.getTokenAddressBySymbol('DGD'),
+        const accounts = await web3Wrapper.getAvailableAddressesAsync();
+        maker = accounts[0];
+        feeRecipient = accounts[1] || accounts[accounts.length - 1];
```
Comment (fabioberger) [inline packages/contracts/test/exchange/helpers.ts]: What is this magic? In which scenarios is `accounts[1]` undefined?

File: packages/contracts/test/exchange/wrapper.ts [hunk 3]
```diff
@@ -38,49 +44,57 @@ contract('Exchange', (accounts: string[]) => {
...
-        [tokenRegistry, exchange] = await Promise.all([TokenRegistry.deployed(), Exchange.deployed()]);
-        exWrapper = new ExchangeWrapper(exchange);
-        const [repAddress, dgdAddress, zrxAddress] = await Promise.all([
-            tokenRegistry.getTokenAddressBySymbol('REP'),
-            tokenRegistry.getTokenAddressBySymbol('DGD'),
-            tokenRegistry.getTokenAddressBySymbol('ZRX'),
+        const accounts = await web3Wrapper.getAvailableAddressesAsync();
+        tokenOwner = maker = accounts[0];
+        taker = accounts[1] || accounts[accounts.length - 1];
+        feeRecipient = accounts[2] || accounts[accounts.length - 1];
```
Comment (fabioberger) [inline packages/contracts/test/exchange/wrapper.ts]: I really dislike these. So one environment is TestRPC where we have 10 accounts. What is the second one? And should these tests work for both the case where `taker` === `feeRecipient` and when not? Sounds error prone to me. @abandeali1

File: packages/contracts/test/multi_sig_with_time_lock_except_remove_auth_addr.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+const PROXY_ABI = artifacts.TokenTransferProxyArtifact.networks[constants.TESTRPC_NETWORK_ID].abi;
+const MUTISIG_WALLET_WITH_TIME_LOCK_EXCEPT_REMOVE_AUTHORIZED_ADDRESS_ABI =
+    artifacts.MultiSigWalletWithTimeLockExceptRemoveAuthorizedAddressArtifact.networks[constants.TESTRPC_NETWORK_ID]
+        .abi;
+
+chaiSetup.configure();
+const expect = chai.expect;
+const web3 = web3Factory.create();
+const web3Wrapper = new Web3Wrapper(web3.currentProvider);
+const blockchainLifecycle = new BlockchainLifecycle(devConstants.RPC_URL);
```
Comment (fabioberger) [inline packages/contracts/test/multi_sig_with_time_lock_except_remove_auth_addr.ts]: Again `devConstants.RPC_URL` is implicitly the same in `web3` instance.

File: packages/contracts/test/multi_sig_with_time_lock_except_remove_auth_addr.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        const accounts = await web3Wrapper.getAvailableAddressesAsync();
+        owners = [accounts[0], accounts[1]];
+        authorizedAddress = `0x${crypto
+            .solSHA3([accounts[0]])
+            .slice(0, 20)
+            .toString('hex')}`;
+        unauthorizedAddress = `0x${crypto
+            .solSHA3([accounts[1]])
+            .slice(0, 20)
+            .toString('hex')}`;
```
Comment (fabioberger) [inline packages/contracts/test/multi_sig_with_time_lock_except_remove_auth_addr.ts]: These transformations should be in a helper function so that they aren't repeated. I know this is legacy code but if you could improve it, that would be awesome.

File: packages/contracts/test/zrx_token.ts [hunk 6]
```diff
@@ -1,60 +1,69 @@
...
+    beforeEach(async () => {
+        await blockchainLifecycle.startAsync();
+    });
+    afterEach(async () => {
+        await blockchainLifecycle.revertAsync();
+    });
     describe('constants', () => {
         it('should have 18 decimals', async () => {
-            const decimals = new BigNumber(await zrx.decimals.call());
+            const decimals = new BigNumber(await zrx.decimals());
```
Comment (fabioberger) [inline packages/contracts/test/zrx_token.ts]: Hm, these methods on the contract wrapper don't have `async` suffixes?

File: packages/contracts/test/zrx_token.ts [hunk 7]
```diff
@@ -73,8 +82,7 @@ contract('ZRXToken', (accounts: string[]) => {
             const receiver = spender;
             const initOwnerBalance = await zeroEx.token.getBalanceAsync(zrxAddress, owner);
             const amountToTransfer = new BigNumber(1);
-            const txHash = await zeroEx.token.transferAsync(zrxAddress, owner, receiver, amountToTransfer);
-            await zeroEx.awaitTransactionMinedAsync(txHash);
```
Comment (fabioberger) [inline packages/contracts/test/zrx_token.ts]: Why is it no longer necessary to await the tx to be mined?

File: packages/deployer/src/utils/contract.ts [hunk 8]
```diff
@@ -28,16 +28,16 @@ export class Contract implements Web3.ContractInstance {
         _.forEach(functionsAbi, (functionAbi: Web3.MethodAbi) => {
             if (functionAbi.constant) {
                 const cbStyleCallFunction = this._contract[functionAbi.name].call;
-                this[functionAbi.name] = {
-                    callAsync: promisify(cbStyleCallFunction, this._contract),
```
Comment (fabioberger) [inline packages/deployer/src/utils/contract.ts]: Why did we get rid of `callAsync` and re-instate the use of `call`? It's still an async call no?

File: packages/types/src/index.ts [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import { BigNumber } from '@0xproject/utils';
+import { BigNumber } from 'bignumber.js';
```
Comment (fabioberger) [inline packages/types/src/index.ts]: Why this change?

File: packages/utils/src/abi_decoder.ts [hunk 10]
```diff
@@ -1,9 +1,10 @@
-import { BigNumber } from '@0xproject/utils';
+import { AbiType, DecodedLogArgs, LogWithDecodedArgs, RawLog, SolidityTypes } from '@0xproject/types';
 import * as _ from 'lodash';
 import * as Web3 from 'web3';
 import * as SolidityCoder from 'web3/lib/solidity/coder';
 
-import { AbiType, ContractEventArgs, DecodedLogArgs, LogWithDecodedArgs, RawLog, SolidityTypes } from '../types';
+// tslint:disable-next-line:no-unused-variable
```
Comment (fabioberger) [inline packages/utils/src/abi_decoder.ts]: Can you add a line explaining why this tslint is being ignored here.

File: packages/contracts/test/exchange/wrapper.ts (hunk 3 above)
Comment (abandeali1) [reply packages/contracts/test/exchange/wrapper.ts]: Yeah you're right, I think a lot of these tests would probably fail if `taker` === `feeRecipient`

File: packages/contracts/test/ether_token.ts [hunk 11]
```diff
@@ -1,43 +1,48 @@
...
+import { constants } from '../util/constants';
 
 import { chaiSetup } from './utils/chai_setup';
-
-const { EtherToken } = new Artifacts(artifacts);
+import { deployer } from './utils/deployer';
 
 chaiSetup.configure();
 const expect = chai.expect;
+const web3 = web3Factory.create();
```
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/contracts/test/ether_token.ts]: We do the same in 0x.js and I don't see how this is not future-proof. Let's not over-engineer before it's needed. I agree that that's kinda more explicit, but I'm generally against adding new params if they could as well be constants inside.

File: packages/contracts/test/exchange/helpers.ts [hunk 12]
```diff
@@ -1,52 +1,68 @@
...
 import { chaiSetup } from '../utils/chai_setup';
+import { deployer } from '../utils/deployer';
 
 chaiSetup.configure();
 const expect = chai.expect;
 
-const { Exchange, TokenRegistry } = new Artifacts(artifacts);
+const web3 = web3Factory.create();
+const web3Wrapper = new Web3Wrapper(web3.currentProvider);
+const blockchainLifecycle = new BlockchainLifecycle(devConstants.RPC_URL);
```
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/contracts/test/exchange/helpers.ts]: Ok, now I see. So should we remove this param here or should we add it to web3Factory?

File: packages/contracts/test/exchange/helpers.ts (hunk 2 above)
Comment (LogvinovLeon) [reply packages/contracts/test/exchange/helpers.ts]: I wish I knew the answer. It was like that before.

File: packages/contracts/test/zrx_token.ts (hunk 6 above)
Comment (LogvinovLeon) [reply packages/contracts/test/zrx_token.ts]: They don't for now, but they will with the upcomming abi-gen PR

File: packages/contracts/test/zrx_token.ts (hunk 7 above)
Comment (LogvinovLeon) [reply packages/contracts/test/zrx_token.ts]: Tests run only against testrpc. On testrpc we have instant mining and that's why we await transactions only if we need a receipt.

File: packages/deployer/src/utils/contract.ts (hunk 8 above)
Comment (LogvinovLeon) [reply packages/deployer/src/utils/contract.ts]: It is. Will fix that in abi-gen PR

File: packages/types/src/index.ts (hunk 9 above)
Comment (LogvinovLeon) [reply packages/types/src/index.ts]: Because here we needed only the type and it helped to break a circular dependency.

File: packages/utils/src/abi_decoder.ts (hunk 10 above)
Comment (LogvinovLeon) [reply packages/utils/src/abi_decoder.ts]: I wish I knew that. Looks like a tslint bug, cause the import is clearly not unused.

Comment (LogvinovLeon) [reply packages/utils/src/abi_decoder.ts]: There are some other linter false-positives that I'm investigating right now
